V1 PreP2P Module Prototype

Olshansk opened the pull request:
# Description

This PR is meant to address https://github.com/pokt-network/pocket/issues/32.

It is a temporary P2P module used to enable the V1 prototype on mainline.

# Details

- Consolidated PocketEvents/NetworkEvents/PocketMessage/NetworkMessage into one proto type
- Implement `pre2p`
    - Add a Network type
    - Supports basic `Send` and `Broadcast` using the golang std library
- Use proper types for address and privateKey in `config.go`
    - Add Unmarshalling of the Private & Public keys from file
- Remove all reference to `node_id` or `NodeId` from the pre-prototype
- Change the address type from `string` to `pcrypto.Address`
- Change the topic type from `string` to `types.PocketTopic`
- Add `protogen_local` to Makefile
- Minor update to .gitignore

## TODOs as a result of the discussion PR
** (will create an issue and add these before submission)
- [ ] **[Small]** Remove `pre2p` altogether
- [ ] **[Medium]** Discuss how to handle (avoid or do) self broadcasts in `pre2p`
- [ ] **[Large]** Discuss how to handle unexpected behavior. Panic? Fail Log? Something else?
- [ ] **[Large]** Add support for dynamic validator sets
- [ ] **[Large]** Proper discussion and design around a genesis file
- [ ] **[Large]** Proper discussion and design around how to access global state (i.e. remove `test_state.go`)

## Testing

Please note that until consensus is integrated into mainline, there is no business logic,
but we can connect to the open port via `telnet` once localnet starts up.

## 1. Pull changes (if not already local):
```
$ git clone git@github.com:pokt-network/pocket.git pocket
$ cd pocket
$ gco -b issue32/prototype/pre2p origin/issue32/prototype/pre2p
```

## [Optional] 2. View available commands:
```
$ make
```

## 3. Clean environment:
```
$ make docker_wipe # Press `y` if prompted
```

### 3.1 Shell 1:
```
$ make protogen_clean
$ make protogen_local
$ make compose_and_watch
```

### 3.2 Shell 2:
```
$ telnet localhost 8081
```

## [Optional] 4.  Foundation for unit tests:
```
$ make mockgen
$ make test_all
```

andrewnguyen22 commented on pre2p/types/validator.go:
```diff
@@ -0,0 +1,40 @@
+
+// TODO(olshansky): Review all the attributes of this struct.
+type Validator struct {
+	NodeId     NodeId   `json:"node_id"`
```
So many todo's haha. 

Similar to above, I'd recommend deprecating nodeId for the public key if possible. I see a TODO to remove it above, so won't harp or block cause of this

Olshansk replied:
Removing NodeId altogether. It'll make the consensus PR a little more complicated but I think it's the right move this early on.

Olshansk replied:
Note that in order to remove NodeId from everywhere, I had to change the Address type from a byte slice to a byte array (i.e. fixed length with size `AddressLen`). Lmk if that's an issue.

andrewnguyen22 replied:
Actually this is a problem

Protobuf only allows []byte not fixed length

So it's either we convert all slices to byte arrays using `copy` or we change address encoding to strings

Olshansk replied:
Discussed this offline with Andrew directly. The decision was:
- Leave `Address` as a slice rather than a string or array for proto compatibility and confirmation to standards
- In the map (where a key cannot be a slice), simply use a string (hex representation of the address) without needing to create an extra type.

andrewnguyen22 commented on pre2p/genesis.go:
```diff
@@ -0,0 +1,70 @@
+// TODO(olshansky): This is an interim genesis structure that will be replaced with a real one. It is the bare minimum for prototyping purposes.
+type P2PGenesis struct {
+	GenesisTime time.Time          `json:"genesis_time"`
+	AppHash     string             `json:"app_hash"`
```
No idea why P2PGenesis needs an appHash... Can this be removed?

Olshansk replied:
The hash of the last block for the V0 blockchain when we do the migration?

andrewnguyen22 replied:
Yes but why for P2P? Consensus might need one.

andrewnguyen22 commented on pre2p/genesis.go:
```diff
@@ -0,0 +1,70 @@
+	if len(genesis.Validators) == 0 {
+		return fmt.Errorf("Genesis must contain at least one validator")
+	}
+
```
If you end up keeping the appHash (not recommended) then we need validation here

Olshansk replied:
tl;dr Validation added.

Added validation of the appHash for the time being, but can be removed depending on the outcome of the thread above.

andrewnguyen22 commented on pre2p/network.go:
```diff
@@ -0,0 +1,89 @@
+			log.Println("Error connecting to peer: ", err)
+			continue
+		}
+		client.Write(data)
```
Need error handling. Don't forget to close the client

Olshansk replied:
Let me know regarding my reasoning above.

andrewnguyen22 commented on pre2p/network.go:
```diff
@@ -0,0 +1,89 @@
+			log.Println("Error connecting to peer: ", err)
+			continue
+		}
+		client.Write(data)
```
Need error handling. Don't forget to close the client

Olshansk replied:
Done.

andrewnguyen22 commented on pre2p/network.go:
```diff
@@ -0,0 +1,89 @@
+	var tcpAddr *net.TCPAddr
+	var err, errDebug error
+
+	tcpAddr, err = net.ResolveTCPAddr("tcp4", fmt.Sprintf("%s:%d", v.Host, v.Port))
```
Move "tcp4" to constant or find standard constant

Olshansk replied:
Using `const NetworkProtocol = "tcp4"` everywhere in the file now.

andrewnguyen22 commented on pre2p/network.go:
```diff
@@ -0,0 +1,89 @@
+
+func (n *network) connectToValidator(nodeId types.NodeId, v *types.Validator) error {
+	var tcpAddr *net.TCPAddr
+	var err, errDebug error
```
I don't understand errDebug. Doesn't look used

Olshansk replied:
tl;dr Removed.

This was from when I was exposing one port for debugging/telemetry and one for the actual communication, but removing the debug/telemetry port in this initial version.

andrewnguyen22 commented on pre2p/handlers.go:
```diff
@@ -0,0 +1,39 @@
+
+	data, err := ioutil.ReadAll(conn)
+	if err != nil {
+		log.Println("Error reading from conn: ", err)
```
Seeing some fmt.Print... and some log.Println

I know we don't yet have a standard, but seems odd to not standardize the usage in this PR

Olshansk commented on pre2p/types/message.go:
```diff
@@ -0,0 +1,8 @@
+
+type NetworkMessage struct {
+	Topic types.EventTopic
+	Data  []byte
```
Resolving this comment in favor of continuing the conversation in the thread above.

Olshansk commented on pre2p/types/proto/p2p.proto:
```diff
@@ -0,0 +1,17 @@
+
+option go_package = "/types/";
+
+enum PocketTopic {
```
I consolidated PocketEvent and NetworkMessage in shared/events.proto.

It's a fair number of changes but simplifies the # of messages overall. Lmk what you think.

Olshansk commented on pre2p/types/network.go:
```diff
@@ -0,0 +1,21 @@
+	"net"
+)
+
+// TODO(olshansky): Try to find a way to remove `NodeId` from the entire codebase altogether; this is a stop-gap prototype measure.
```
Removing NodeId from broadcast.

derrandz commented on .gitignore:
```diff
@@ -24,9 +24,11 @@ vendor/
 
 # Proto generated files
-types/
+*.pb.go
```
Good catch

derrandz commented on p2p/pre2p/handlers.go:
```diff
@@ -0,0 +1,38 @@
+	}
+
+	event := types.PocketEvent{
+		Module: types.SourceModule_P2P_MODULE,
```
I wonder if a `Module` property is necessary on an event. A topic should imply enough context. However, I see how useful can this be for debugging purposes.

Noting for later, this can benefit from some thought.

Olshansk replied:
Removing. We can always add it back later if need be.

derrandz commented on p2p/pre2p/module.go:
```diff
@@ -49,10 +61,57 @@ func (m *p2pModule) GetBus() modules.Bus {
-func (m *p2pModule) Send(addr string, msg *anypb.Any, topic string) error {
-	panic("Send not implemented")
+func (m *p2pModule) GetAddrBook() []*pre2ptypes.NetworkPeer {
+	return m.network.GetAddrBook()
```
Good looking file.

derrandz commented on p2p/pre2p/network.go:
```diff
@@ -0,0 +1,98 @@
+
+func (n *network) NetworkSend(data []byte, node pcrypto.Address) error {
+	for _, peer := range n.AddrBook {
+		// TODO(team): If the address book is a map instead of a list, we wouldn't have to do this loop.
```
This contains the answer to the question at [line 35](https://github.com/pokt-network/pocket/pull/33/files#diff-4c195a5d75f98cf6df9e78d6fc0766ada49a8a40af22a3c4385d3c4becc008dcR35) as a list containing only neighbors would be a self-filtering mechanism to prevent self-broadcast, as long as other peers do not re-broadcast to not fall into infinite broadcast.

derrandz replied:
I suppose what I implied since Pre2P is always going to be used in controlled environments that we establish the expectation that the neighbors list will never contain own address, making of it a natural self-filtering mechanism. That's why making it a list and not a map will be much more useful.

Olshansk replied:
Makes sense. We'll just need to validate that when the static list / addrbook is created /configured, self is not in there.

Added a comment in the code.

derrandz commented on shared/handler.go:
```diff
@@ -6,14 +6,12 @@ import (
-		node.GetBus().GetConsensusModule().HandleMessage(event.MessageData)
+func (node *Node) handleEvent(event *types.PocketEvent) error {
+	switch event.Topic {
+	case types.PocketTopic_CONSENSUS_MESSAGE_TOPIC:
```
As a comment, I think I prefer shorter names. `types.EVENT_TOPIC_CONSENSUS_MESSAGE`. There is quite the redundancy in `PocketTopic_...._TOPIC`.

Olshansk replied:
This is a proto generated type from an enum so outside of our control.

derrandz replied:
Yeah correct...

derrandz commented on shared/handler.go:
```diff
@@ -6,14 +6,12 @@ import (
 	default:
-		// TODO(discuss): Should we panic here?
-		log.Printf("[WARN] Unsupported event: %s \n", event.PocketTopic)
+		log.Printf("Unsupported event: %s \n", event.Topic)
```
We should think about handling undefined behavior, including panic recovery and unexpected behaviors such as these. I am talking in general as P2P also lacks this aspect (_intentionally deferred to be thought about thoroughly, perhaps collectively?_)

Olshansk replied:
Adding a top-level TODO at the PR level. Would you want to add it to a future protocol hour and lead the discussion?

derrandz replied:
Good suggestion Shans, will do.

Olshansk replied:
Thanks Zanzan

derrandz commented on shared/modules/p2p_module.go:
```diff
@@ -1,18 +1,13 @@
 	Module
-	BroadcastMessage(msg *anypb.Any, topic string) error  // TODO(derrandz): get rid of topic
-	Send(addr string, msg *anypb.Any, topic string) error // TODO(derrandz): get rid of topic
+	BroadcastMessage(msg *anypb.Any, topic types.PocketTopic) error           // TODO(derrandz): get rid of topic
```
I think these signatures are better.

derrandz commented on shared/modules/pocket_bus_module.go:
```diff
@@ -4,14 +4,14 @@ import (
-	PublishEventToBus(e *types.Event)
-	GetBusEvent() *types.Event
+	PublishEventToBus(e *types.PocketEvent)
+	GetBusEvent() *types.PocketEvent
```
(_Just a comment_): I remember a short conversation about idiomatic go. I noticed also that in the generated protobuff golang code, they do not seem to respect the idiomatic way of naming getters. Overall, it's simpler to read `GetBusEvent` than `BusEvent`. Still, I want to hear other opinions, particularly yours.

Olshansk replied:
I have no strong opinions either way. I think either approach enables us to do what we need.

derrandz commented on shared/types/proto/events.proto:
```diff
@@ -0,0 +1,33 @@
+
+option go_package = "/types/";
+
+enum SourceModule {
```
I think it's more clean to name this something like: `Concern' or 'Scope' than 'Module'. Since a naming like "module" is somewhat meta (_as in meta-programming_) and reverts the attention off of the actual business of the code into the code structure itself.

I think `SCOPE` is more fitting.

Olshansk replied:
Removing altogether until we need it again.

derrandz reviewed (approved):
Overall LGTM. Minor comments to strike conversations, with relatively strong preferential opinions about naming.

Thanks for the good work good sir 👍🏻

Olshansk commented:
@andrewnguyen22 I believe I've addressed most of the issues. Could you take another look at this to see if everything 👀 good? If so, we can submit it for @derrandz to work on top of in his branch.
